### ISSUE TITLE
ClassCastException Vermögensaufstellung Diagramm

### ISSUE THREAD

**Issue by phiradd** [buchen/portfolio#1833]
When clicking on Vermögensaufstellung -> Diagramm instead of the Chart I get a ClassCastException, see below.

Version: 0.49.2 (Nov 2020)
Platform: win32, x86_64
Java: 11.0.4+11-LTS, Azul Systems, Inc.

```
class name.abuchen.portfolio.model.Portfolio cannot be cast to class name.abuchen.portfolio.model.Account (name.abuchen.portfolio.model.Portfolio and name.abuchen.portfolio.model.Account are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @34d52ecd)

java.lang.ClassCastException: class name.abuchen.portfolio.model.Portfolio cannot be cast to class name.abuchen.portfolio.model.Account (name.abuchen.portfolio.model.Portfolio and name.abuchen.portfolio.model.Account are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @34d52ecd)

	at name.abuchen.portfolio.snapshot.filter.ClientClassificationFilter.adaptAccountTransactions(ClientClassificationFilter.java:334)

	at name.abuchen.portfolio.snapshot.filter.ClientClassificationFilter.filter(ClientClassificationFilter.java:131)

	at name.abuchen.portfolio.snapshot.PerformanceIndex.forClassification(PerformanceIndex.java:102)

	at name.abuchen.portfolio.ui.views.dataseries.DataSeriesCache.calculate(DataSeriesCache.java:124)

	at name.abuchen.portfolio.ui.views.dataseries.DataSeriesCache.lookup(DataSeriesCache.java:71)

	at name.abuchen.portfolio.ui.views.dataseries.StatementOfAssetsSeriesBuilder.build(StatementOfAssetsSeriesBuilder.java:21)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.lambda$3(StatementOfAssetsHistoryView.java:116)

	at java.base/java.lang.Iterable.forEach(Unknown Source)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.createBody(StatementOfAssetsHistoryView.java:116)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.createBody(StatementOfAssetsHistoryView.java:1)

	at name.abuchen.portfolio.ui.editor.AbstractFinanceView.createViewControl(AbstractFinanceView.java:127)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.createView(PortfolioPart.java:525)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.activateView(PortfolioPart.java:495)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.activateView(PortfolioPart.java:483)

	at name.abuchen.portfolio.ui.editor.ClientEditorSidebar$1.select(ClientEditorSidebar.java:70)

	at name.abuchen.portfolio.ui.editor.ClientEditorSidebar$1.select(ClientEditorSidebar.java:1)

	at name.abuchen.portfolio.ui.editor.Sidebar$Entry.handleMouseDown(Sidebar.java:419)

	at org.eclipse.swt.events.MouseListener$2.mouseDown(MouseListener.java:96)

	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:196)

	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)

	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4213)

	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1037)

	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4030)

	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3630)

	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1158)

	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)

	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1047)

	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)

	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.start(E4Application.java:166)

	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)

	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)

	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)

	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)

	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)

	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)

	at java.base/java.lang.reflect.Method.invoke(Unknown Source)

	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)

	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)

	at org.eclipse.equinox.launcher.Main.run(Main.java:1447)

```

**Comment by buchen** [comment]
> When clicking on Vermögensaufstellung -> Diagramm instead of the Chart I get a ClassCastException


Thanks for the error report. Very strange - it appears like there is a cash transfer which goes to a securities account - something that should not possible.

Probably the XML got corrupted through some other action - and now it breaks at this place.

Would you be willing to share your XML file via email to *portfolio dot performance dot help at gmail dot com*? Feel free to delete some transactions, but do not delete any cash transfers and make sure the problem still occurs. Thanks

**Comment by phiradd** [comment]
I sent you an email with the xml-file. Please let me know if you haven't received it or need any further info.

**Comment by buchen** [comment]
Hi @phiradd - somehow I can't seem to find your email in my inbox. 🤔 I also check my spam folder. Can you try again: *portfolio.performance.help (at) gmail.com* Sorry if I missed it - I am getting lots and lots of emails

**Comment by phiradd** [comment]
Hi @buchen - I resent it just now. I have not gotten any maildaemon errors or anything, so it should have reached you. Please let me know if not. I could try an alternative mail server of mine.

**Comment by buchen** [comment]
Thanks. Can you run *File -> Consistency Check*. That should "fix" the problem. But there will be a new security in the list and you have to decide what to do with it.

<img width="363" alt="Bildschirmfoto 2020-11-29 um 17 11 40" src="https://user-images.githubusercontent.com/587976/100547245-f47f4800-3265-11eb-9f31-47497d888a28.png">


I still do not know how this situation is created though...

**Comment by phiradd** [comment]
Thank you very much. Indeed, this has solved the problem.

**Comment by r-eis** [comment]
Seems like I have the same problem, but the consistency check did not find anything. 
It happens if I add a certain account to an asset class that is selected in the portfolio diagram.
It does not happen if I add that account directly to the portfolio diagram, but probably you saw that from the stack trace.

Version: 0.51.2 (März 2021)
Platform: win32, x86_64
Java: 11.0.4+11-LTS, Azul Systems, Inc.

Stack trace:

`class name.abuchen.portfolio.model.Portfolio cannot be cast to class name.abuchen.portfolio.model.Account (name.abuchen.portfolio.model.Portfolio and name.abuchen.portfolio.model.Account are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @6edf29c1)

java.lang.ClassCastException: class name.abuchen.portfolio.model.Portfolio cannot be cast to class name.abuchen.portfolio.model.Account (name.abuchen.portfolio.model.Portfolio and name.abuchen.portfolio.model.Account are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @6edf29c1)

	at name.abuchen.portfolio.snapshot.filter.ClientClassificationFilter.adaptAccountTransactions(ClientClassificationFilter.java:344)

	at name.abuchen.portfolio.snapshot.filter.ClientClassificationFilter.filter(ClientClassificationFilter.java:135)

	at name.abuchen.portfolio.snapshot.PerformanceIndex.forClassification(PerformanceIndex.java:102)

	at name.abuchen.portfolio.ui.views.dataseries.DataSeriesCache.calculate(DataSeriesCache.java:124)

	at name.abuchen.portfolio.ui.views.dataseries.DataSeriesCache.lookup(DataSeriesCache.java:71)

	at name.abuchen.portfolio.ui.views.dataseries.StatementOfAssetsSeriesBuilder.build(StatementOfAssetsSeriesBuilder.java:21)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.lambda$3(StatementOfAssetsHistoryView.java:116)

	at java.base/java.lang.Iterable.forEach(Unknown Source)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.createBody(StatementOfAssetsHistoryView.java:116)

	at name.abuchen.portfolio.ui.views.StatementOfAssetsHistoryView.createBody(StatementOfAssetsHistoryView.java:1)

	at name.abuchen.portfolio.ui.editor.AbstractFinanceView.createViewControl(AbstractFinanceView.java:127)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.createView(PortfolioPart.java:533)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.activateView(PortfolioPart.java:503)

	at name.abuchen.portfolio.ui.editor.PortfolioPart.activateView(PortfolioPart.java:491)

	at name.abuchen.portfolio.ui.editor.ClientEditorSidebar$1.select(ClientEditorSidebar.java:70)

	at name.abuchen.portfolio.ui.editor.ClientEditorSidebar$1.select(ClientEditorSidebar.java:1)

	at name.abuchen.portfolio.ui.editor.Sidebar$Entry.handleMouseDown(Sidebar.java:419)

	at org.eclipse.swt.events.MouseListener$2.mouseDown(MouseListener.java:96)

	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:196)

	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)

	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4209)

	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1037)

	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4026)

	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3626)

	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1157)

	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)

	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)

	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)

	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.start(E4Application.java:166)

	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)

	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)

	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)

	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)

	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)

	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)

	at java.base/java.lang.reflect.Method.invoke(Unknown Source)

	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)

	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)

	at org.eclipse.equinox.launcher.Main.run(Main.java:1461)


`

**Comment by r-eis** [comment]
I identfied the reason for this.
It was an entry booking of type "Umbuchung" whose counter account was my portfolio account.
I guess, this kind of booking should never exist.
The booking was created by the PDF importer for the DKB bank for a corporate bond refund. 

I also attached the extracted and anonymized text of the PDF. Maybe that helps with debugging.
[pdf-content.txt](https://github.com/buchen/portfolio/files/6178077/pdf-content.txt)
